TSDB Data stream integration prototype 1

martijnvg opened the pull request:
Changes:
* Added notion of a type to data streams and introduced two types:
  `DEFAULT` and `TSDB`.
* If composable template has a data stream template definition with
  type TSDB then the `index.time_series.start_time` and
  `index.time_series.end_time` index settings get applied to the
  first backing index. Start time will be set to current time - 5
  minutes and end time will be set to current time + look ahead time.
* When rolling over a data stream with type TSDB then the new
  backing index will be set with `index.time_series.start_time` and
  `index.time_series.end_time` index settings. The start time will be
  set to the end time of the previous latest backing index and the end
  time will be set to start time + look ahead time + poll interval.
* The IndexNameExpressionResolver is adjusted to in case when a
  TSDB data stream is resolved to parse the `@timestamp` field of
  the document being ingested and use that to select the right
  backing index.
* Added a new service that updates that periodically updates the
  `index.time_series.end_time` index setting of the latest backing
  index of all data streams that have type TSDB to now + look_ahead_time + poll_interval.


Sample snippets to play around with prototype:

```
PUT _index_template/my-template
{
  "index_patterns": ["test-*"],
  "data_stream": {
      "type": "tsdb"
  },
  "template": {
    "settings": {
        "index":{
            "mode": "time_series",
            "routing_path": "uid"
        }
    },
    "mappings": {
        "properties": {
            "uid": {
                "type": "keyword",
                "time_series_dimension": true
            }
        }
    }
  }
}

POST test-data1/_doc
{
  "@timestamp": "2021-11-29T15:56:50Z",
  "uid": "_uid"
}

GET /_data_stream

POST /test-data1/_rollover
```

martijnvg committed:
TSDB Data stream integration prototype 1

Changes:
* Added notion of a type to data streams and introduced two types:
  `DEFAULT` and `TSDB`.
* If composable template has a data stream template definition with
  type TSDB then the `index.time_series.start_time` and
  `index.time_series.end_time` index settings get applied to the
  first backing index. Start time will be set to current time - 5
  minutes and end time will be set to current time + look ahead time.
* When rolling over a data stream with type TSDB then the new
  backing index will be set with `index.time_series.start_time` and
  `index.time_series.end_time` index settings. The start time will be
  set to the end time of the previous latest backing index and the end
  time will be set to start time + look ahead time + poll interval.
* The IndexNameExpressionResolver is adjusted to in case when a
  TSDB data stream is resolved to parse the `@timestamp` field of
  the document being ingested and use that to select the right
  backing index.
* Added a new service that updates that periodically updates the
  `index.time_series.end_time` index setting of the latest backing
  index of all data streams that have type TSDB.

martijnvg committed:
Simplify by using INER only to resolve IndexAbstraction and then
let TransportBulkAction to resolve concrete write index from IndexAbstraction

martijnvg committed:
fix number of tests and also check for system access

martijnvg committed:
Merge remote-tracking branch 'es/master' into tsdb_data_streams_prototype1

martijnvg committed:
Isolate index cannot be created check and check that earlier in the execution

martijnvg committed:
fix docs test

martijnvg committed:
Also check if alias has write index in the new INER#resolveWriteIndexAbstraction() method

martijnvg committed:
make type optional for bwc reasons

martijnvg committed:
removed unused code

martijnvg committed:
Always update end_time each poll interval to now + look_ahead_time.

martijnvg committed:
Fixed issues with creating correct temporal boundaries

martijnvg committed:
Added time series start and end time to get data stream api response for tsdb data streams

martijnvg committed:
spotless

nik9000 commented:
> * Start time will be set to current time - 5
>       minutes and end time will be set to current time + look ahead time.

I wonder if the first index should have the start time default to 1970 or something.

nik9000 commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -730,40 +729,19 @@ void executeBulk(
+        IndexAbstraction resolveIfAbsent(DocWriteRequest<?> request) {
+            return indexAbstractions.computeIfAbsent(
+                request.index(),
+                key -> indexNameExpressionResolver.resolveWriteIndexAbstraction(state, request)
```
Will this cache the index name to the index resolved for the request? I was thinking it'd have to be something like:
```
return indexAbstractions.computeIfAbsent(request.index()).resolveWriteIndexAbstraction(request);
```

martijnvg replied:
The `state` is needed for the `resolveWriteIndexAbstraction(...)` method to be able to resolve specified index expression on the request to an index, alias or data stream. We use the same cluster state instance throughout the bulk request handling here.

nik9000 commented on server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:
```diff
@@ -170,6 +201,37 @@ public Index getWriteIndex() {
+            return getWriteIndex();
+        }
+
+        // TODO: this really needs be parsed in a streaming manner:
```
`XContentParser` now takes a filter. So you can build it with the filter and then do

```
try (XContentParser parser = TS_EXTRACT_CONFIG.parser(request.type(), request.source())) {
  expect(START_OBJECT, parser.next());
  expect(FIELD_NAME, parser.next());
  expect(VALUE, parser.next());
  return DateFieldMaper.DEFAULT_DATE_TIME_FORMATTER.parseMillis(parser.text());
}

nik9000 commented on server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:
```diff
@@ -170,6 +201,37 @@ public Index getWriteIndex() {
+
+        // TODO: this really needs be parsed in a streaming manner:
+        String timestampAsString = (String) ((IndexRequest) request).sourceAsMap().get("@timestamp");
+        long timestamp = DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER.parseMillis(timestampAsString);
```
We should probably throw an error in the mapping if the field doesn't have the default formatter.

Hmmmmmm.......... What about nanos?

martijnvg replied:
I think we should check mapping in the cluster state and try to detect what formatter to use here.
This should also allow this code to parse nanos. I will add an TODO for this.

nik9000 commented on server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:
```diff
@@ -170,6 +201,37 @@ public Index getWriteIndex() {
     }
 
+    public Index getWriteIndex(DocWriteRequest<?> request, Metadata metadata) {
+        if (type != Type.TSDB) {
```
Maybe a bad idea - could we check that the `getWriteIndex().mode() == TSDB`?

I was sort of hoping that `IndexRouting` could handle this kind of redirection somehow because it "feels right". We already have a specialization of it for tsdb. And, it's, like, testable without building the whole world.

nik9000 replied:
Also, maybe it'd be nice not to have to add a type to data streams - they just know their type based on the type of the write index.

martijnvg replied:
> Maybe a bad idea - could we check that the getWriteIndex().mode() == TSDB?

👍 

> Also, maybe it'd be nice not to have to add a type to data streams - they just know their type based on the type of the write index.

I will look into whether we can detect how to select to index to write into. I agree this would be nice and cleaner. The `DataStream` class doesn't have access to index settings or other index related metadata, so I need to think on this.

imotov commented on server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:
```diff
@@ -562,7 +637,18 @@ public boolean equals(Object o) {
+        return Objects.hash(name, type, timeStampField, indices, generation, metadata, hidden, replicated, allowCustomRouting);
+    }
+
+    public enum Type {
```
Can we reuse IndexMode here? If this is not possible or impractical, I think we should at least reuse the same names here and in IndexMode.

martijnvg replied:
> Can we reuse IndexMode here?

I will look into this as well and conjunction with Nik's comment about not have to use a type at all.

> I think we should at least reuse the same names here and in IndexMode.

Agreed, at least the naming should be consistent between the two enums.

imotov reviewed (commented):
Thanks! Looks like a good start!

martijnvg commented:
> I wonder if the first index should have the start time default to 1970 or something.

I kind of think this makes sense. It is kind of open bounded to one end, but that shouldn't matter. As soon as rollover conditions allow it, the data stream will be rolled over and the next index's temporal boundaries make more sense to the data that ends up being indexed.

martijnvg committed:
Merge remote-tracking branch 'es/master' into tsdb_data_streams_prototype1

martijnvg committed:
Changes:
* Rename enum values
* Change first backing's TIME_SERIES_START_TIME index setting to always be: 1970-01-01T00:00:00
* Tweaked tsdb yaml test
* Added hack to not check for existence of TIME_SERIES_START_TIME and TIME_SERIES_START_TIME index settings.

martijnvg committed:
Merge remote-tracking branch 'es/master' into tsdb_data_streams_prototype1

martijnvg committed:
fix docs tests

martijnvg committed:
adjust yaml test

martijnvg committed:
tweaked hlrc test

martijnvg committed:
streaming parsing of @timestamp field

martijnvg committed:
Merge remote-tracking branch 'es/master' into tsdb_data_streams_prototype1

martijnvg committed:
Stop using data stream type to trigger tsdb behaviour.

martijnvg commented:
@imotov and @nik9000 I've removed the usages of data stream type in all places via 5a4625b8e07366953591a8df332e9ee2868ab591. 

This means data stream type no longer needs to be used in the template and it looks like this now:

```
PUT _index_template/my-template
{
  "index_patterns": ["test-*"],
  "data_stream": {},
  "template": {
    "settings": {
        "index":{
            "mode": "time_series",
            "routing_path": "uid"
        }
    },
    "mappings": {
        "properties": {
            "uid": {
                "type": "keyword",
                "time_series_dimension": true
            }
        }
    }
  }
}
```

I think this is better? I've not removed the notion of a data stream type here, but I will just leave it here in this prototype and not include it in followup PRs.

martijnvg committed:
Forgot one data stream type usage

martijnvg committed:
fix npe

martijnvg committed:
spotless

martijnvg committed:
Merge remote-tracking branch 'es/master' into tsdb_data_streams_prototype1

martijnvg committed:
re-enable start_time and end_time validation and
when validating composable index templates automatically add start_time and end_time setting if
a template has a data stream definition and index.mode=time_series. This mimics what
MetadataCreateDataStreamService is doing.

martijnvg committed:
spotless

martijnvg commented:
Closing as this is a prototype and this will be implemented in separate PRs.
